lua/lspconfig/server_configurations/omnisharp.lua

```diff
@@ -42,37 +43,44 @@ return {
       return util.root_pattern '*.sln'(fname) or util.root_pattern '*.csproj'(fname)
     end,
     on_new_config = function(new_config, new_root_dir)
+      -- store the initial value of `cmd` for later use
+      if not cmd_init then
+        cmd_init = {}
+        for _, v in pairs(new_config.cmd) do
+          table.insert(cmd_init, v)
+        end
+      end
+      -- copy the fields of `cmd_init` into `new_config.cmd`
+      new_config.cmd = {}
+      for _, v in pairs(cmd_init) do
+        table.insert(new_config.cmd, v)
+      end
+      -- add hard-coded command arguments
       table.insert(new_config.cmd, '-z') -- https://github.com/OmniSharp/omnisharp-vscode/pull/4300
       vim.list_extend(new_config.cmd, { '-s', new_root_dir })
       vim.list_extend(new_config.cmd, { '--hostPID', tostring(vim.fn.getpid()) })
       table.insert(new_config.cmd, 'DotNet:enablePackageRestore=false')
       vim.list_extend(new_config.cmd, { '--encoding', 'utf-8' })
       table.insert(new_config.cmd, '--languageserver')
-
+      -- add configuration-dependent command arguments
       if new_config.enable_editorconfig_support then
         table.insert(new_config.cmd, 'FormattingOptions:EnableEditorConfigSupport=true')
       end
-
       if new_config.organize_imports_on_format then
         table.insert(new_config.cmd, 'FormattingOptions:OrganizeImports=true')
       end
-
       if new_config.enable_ms_build_load_projects_on_demand then
         table.insert(new_config.cmd, 'MsBuild:LoadProjectsOnDemand=true')
       end
-
       if new_config.enable_roslyn_analyzers then
         table.insert(new_config.cmd, 'RoslynExtensionsOptions:EnableAnalyzersSupport=true')
       end
-
       if new_config.enable_import_completion then
         table.insert(new_config.cmd, 'RoslynExtensionsOptions:EnableImportCompletion=true')
       end
-
       if new_config.sdk_include_prereleases then
         table.insert(new_config.cmd, 'Sdk:IncludePrereleases=true')
       end
-
       if new_config.analyze_open_documents_only then
         table.insert(new_config.cmd, 'RoslynExtensionsOptions:AnalyzeOpenDocumentsOnly=true')
       end
```

Comment:
Might as well also add this (see 2nd point [here](https://github.com/neovim/nvim-lspconfig/issues/2392#issuecomment-1610129809))?

```suggestion

      new_config.capabilities = vim.deepcopy(new_config.capabilities)
      new_config.capabilities.workspace.workspaceFolders = false -- https://github.com/OmniSharp/omnisharp-roslyn/issues/909
      end
```

Reply:
Thanks, I'll take a look at this soon.

Reply:
Oh, very nice! And it seems that this change also renders the following line&mdash;and function parameter `new_root_dir`&mdash;unnecessary. Would you agree?

``` lua
vim.list_extend(new_config.cmd, { '-s', new_root_dir })
```

I've tested the code with the above line commented out, and it appears to work correctly by starting up a second OmniSharp instance when needed. And it consistently ties new buffers to the appropriate OmniSharp instances.

Reply:
I guess it's not needed:

```
-s|--source              Solution or directory for OmniSharp to point at (defaults to current directory).
```